embassy-mspm0/src/watchdog.rs

```diff
@@ -0,0 +1,341 @@
+pub enum ClosedWindowPercentage {
+    // window period is not used
+    Zero,
+    // 12.5% percents
+    Twelve,
+    // 18.75% percents
+    Eighteen,
+    // 25% percents
+    TwentyFive,
+    // 50% percents
+    Fifty,
+    // 75% percents
+    SeventyFive,
+    // 81.25% percents
+    EightyOne,
+    // 87.5% percents
+    EightySeven,
+}
+
+impl ClosedWindowPercentage {
+    fn get_native_size(self) -> vals::Window {
+        match self {
+            Self::Zero => vals::Window::SIZE_0,
+            Self::Twelve => vals::Window::SIZE_12,
+            Self::Eighteen => vals::Window::SIZE_18,
+            Self::TwentyFive => vals::Window::SIZE_25,
+            Self::Fifty => vals::Window::SIZE_50,
+            Self::SeventyFive => vals::Window::SIZE_75,
+            Self::EightyOne => vals::Window::SIZE_81,
+            Self::EightySeven => vals::Window::SIZE_87,
+        }
+    }
+}
+
+#[non_exhaustive]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+/// Watchdog Config
+pub struct Config {
+    /// Watchdog timeout
+    pub timeout: Timeout,
+
+    /// closed window percentage
+    pub closed_window: ClosedWindowPercentage,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            timeout: Timeout::Sec1,
+            closed_window: ClosedWindowPercentage::Zero,
+        }
+    }
+}
+
+pub struct Watchdog<'d, T: Instance> {
+    wdt: PhantomData<&'d mut T>,
+}
+
+impl<'d, T: Instance> Watchdog<'d, T> {
+    /// Watchdog initialization.
+    pub fn new(_instance: Peri<'d, T>, config: Config) -> Self {
+        // Init power for watchdog
+        T::REGS.gprcm(0).rstctl().write(|w| {
+            w.set_resetstkyclr(true);
+            w.set_resetassert(true);
+            w.set_key(vals::ResetKey::KEY);
+        });
+
+        // Enable power for watchdog
+        T::REGS.gprcm(0).pwren().write(|w| {
+            w.set_enable(true);
+            w.set_key(vals::PwrenKey::KEY);
+        });
+
+        // init delay, 16 cycles
+        cortex_m::asm::delay(16);
+
+        //init watchdog
+        T::REGS.wwdtctl0().write(|w| {
+            w.set_clkdiv(config.timeout.get_clkdiv());
+            w.set_per(config.timeout.get_period());
+            w.set_mode(vals::Mode::WINDOW);
+            w.set_window0(config.closed_window.get_native_size());
+            w.set_window1(vals::Window::SIZE_0);
+            w.set_key(vals::Wwdtctl0Key::KEY);
+        });
+
+        // Set Window0 as active window
+        T::REGS.wwdtctl1().write(|w| {
+            w.set_winsel(vals::Winsel::WIN0);
+            w.set_key(vals::Wwdtctl1Key::KEY);
+        });
+
+        critical_section::with(|_| {
+            // make sure watchdog triggers BOOTRST
+            pac::SYSCTL.systemcfg().write(|w| {
+                w.set_wwdtlp0rstdis(false);
+
+                #[cfg(any(mspm0g110x, mspm0g150x, mspm0g151x, mspm0g310x, mspm0g350x, mspm0g351x))]
+                w.set_wwdtlp1rstdis(false);
```

Comment:
We should not unconditionally write this the wwdt 1 field. If we aren't using WWDT1, we technically do not own its fields or this bit of sysctl. This also lets the user do raw pac access on wwdt if they wanted. 

If we somehow get a 3rd wwdt in the future then we can change that later.

Reply:
fixed